File: dynabuffers-java/src/main/kotlin/dynabuffers/TypeValidator.kt
package dynabuffers

import dynabuffers.antlr.DynabuffersParser
import dynabuffers.exception.DynabuffersException
import org.antlr.v4.runtime.ParserRuleContext
import java.lang.StringBuilder

class TypeValidator {
    private val usedDataTypes = mutableSetOf<String>()
    private val availableDataTypes = mutableSetOf<String>()

    fun validate() {
        val invalidDatatypes = usedDataTypes.minus(availableDataTypes)
        if (invalidDatatypes.isNotEmpty()) throw DynabuffersException("The datatype(s) $invalidDatatypes are neither built in datatypes nor class, union or enum references.")
    }

    fun addAvailableType(ctx: ParserRuleContext) {
        val name = ctx.textAt(1)
        val typeIdentifier = getNamespacePrefix(ctx).append(name).toString()
        availableDataTypes.add(typeIdentifier)
    }

    fun addUsedType(ctx: ParserRuleContext) {
        val name = ctx.text
        val typeIdentifier = getNamespacePrefix(ctx).append(name).toString()
        usedDataTypes.add(typeIdentifier)
    }

    private fun getNamespacePrefix(ctx: ParserRuleContext): StringBuilder {
        val stringBuilder = StringBuilder()
        var context = ctx
        while (context.getParent() !is DynabuffersParser.CompilationContext) {
            context = context.getParent()
            if (context is DynabuffersParser.NamespaceTypeContext)
                stringBuilder.insert(0, "${context.textAt(1)}.")
        }
        return stringBuilder
    }

    private fun ParserRuleContext.textAt(index: Int) = this.getChild(index).text
}

File: dynabuffers-python/dynabuffers/TypeValidator.py
from typing import Union

from antlr4 import ParserRuleContext

from dynabuffers.antlr.DynabuffersParser import DynabuffersParser


class DynabuffersTypeValidator:
    def __init__(self):
        self.visited_data_types = set()
        self.available_data_types = set()

    def validate(self):
        invalid_datatypes = self.visited_data_types - self.available_data_types
        if len(invalid_datatypes) > 0:
            raise ValueError(
                "The datatype(s) {} used in the schema are neither built-in datatypes nor class, union or enum "
                "references.".format(invalid_datatypes))

    def add_available_type(self, ctx: Union[
        DynabuffersParser.ClassTypeContext, DynabuffersParser.UnionTypeContext, DynabuffersParser.EnumTypeContext]):
        name = ctx.getChild(1).getText()
        prefix = self.__get_namespace_prefix(ctx)
        self.available_data_types.add(prefix + name)

    def add_used_type(self, ctx: DynabuffersParser.DataTypeContext):
        name = ctx.getText()
        prefix = self.__get_namespace_prefix(ctx)
        self.visited_data_types.add(prefix + name)

    def __get_namespace_prefix(self, ctx: ParserRuleContext):
        prefix = ""
        while not isinstance(ctx.parentCtx, DynabuffersParser.CompilationContext):
            ctx = ctx.parentCtx
            if isinstance(ctx, DynabuffersParser.NamespaceTypeContext):
                prefix = "{}.{}".format(ctx.getText(), prefix)
        return prefix
